Make delete actually stop resources by default.

brendandburns opened the pull request:
Refactor for shared code.

@bgrant0607 @ghodss @smarterclayton @eparis 

This is a breaking change for `kubectl delete ...` behavior by default.

brendandburns commented:
@deads2k too for good measure ;)

brendandburns committed:
Make delete actually stop resources by default.
Refactor for shared code.

bgrant0607 commented:
And cc @jlowdermilk 

We could put a deprecated message on the stop command, like we did for rollingupdate.

deads2k commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -95,8 +96,37 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+	err := r.Visit(func(info *resource.Info) error {
+		found++
+		reaper, err := f.Reaper(info.Mapping)
+		cmdutil.CheckErr(err)
```
Probably don't want to `fatal` in here.  Can you return the error instead?

brendandburns replied:
done.

deads2k commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -95,8 +96,37 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+
+func ReapResult(r *resource.Result, f *cmdutil.Factory, out io.Writer) error {
+	found := 0
+	err := r.Visit(func(info *resource.Info) error {
```
Looks like `DeleteResult` is ignoring `NotFound` errors. Should this do the same?  I noticed that some`Reapers` are aggressively getting objects before deletion.

smarterclayton replied:
Delete must ignore not found, since it can take duplicate arguments.  Reapers need to tolerate not found as well.

> On Apr 23, 2015, at 8:00 AM, David Eads <notifications@github.com> wrote:
> 
> In pkg/kubectl/cmd/delete.go:
> 
> > @@ -95,8 +96,37 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
> >  		return err
> >  	}
> >  
> > +	// By default use a reaper to delete all related resources.
> > +	if cmdutil.GetFlagBool(cmd, "recursive") {
> > +		return ReapResult(r, f, out)
> > +	}
> > +	return DeleteResult(r, out)
> > +}
> > +
> > +func ReapResult(r *resource.Result, f *cmdutil.Factory, out io.Writer) error {
> > +	found := 0
> > +	err := r.Visit(func(info *resource.Info) error {
> Looks like DeleteResult is ignoring NotFound errors. Should this do the same? I noticed that someReapers are aggressively getting objects before deletion.
> 
> —
> Reply to this email directly or view it on GitHub.
>

brendandburns replied:
done.

brendandburns commented:
Comments addressed.  I'm not convinced we should deprecate ```kubectl stop ...``` I think there is value in ```stop``` as a counter-point to ```run```

We should alias them together, though.

thockin commented:
Without reading the PR, SGTM.  How do we do a "delete and abandon" ?


On Thu, Apr 23, 2015 at 3:01 PM, Brendan Burns <notifications@github.com>
wrote:

> Comments addressed. I'm not convinced we should deprecate kubectl stop ...
> I think there is value in stop as a counter-point to run
>
> We should alias them together, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7210#issuecomment-95731294>
> .
>

bgrant0607 commented:
Delete and abandon is delete --recursive=false.

I'm fine with aliasing stop and delete, though we should document that: #5370.

jlowdermilk commented:
LGTM. +1 to keeping stop as an alias for delete --recursive=true

liggitt commented:
What happens if you `delete` something which doesn't support `stop` without adding `--recursive=false`? I like having the recursive option, but I'm not sure it should be the default

ghodss commented on pkg/kubectl/cmd/stop.go:
```diff
@@ -86,3 +63,24 @@ func NewCmdStop(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	return cmd
 }
+
+func DoStop(f *cmdutil.Factory, cmd *cobra.Command, args []string, filenames util.StringList, out io.Writer) error {
```
Nit... I think it's usually Run*, so RunStop.

brendandburns replied:
done.

ghodss commented:
Is `recursive` the right word here? Obviously it fits for RC's but doesn't this indicate whether to wait for a graceful stop of pods too? (Where `recursive` would be less appropriate.) Or does it not apply to pods?

ghodss commented:
If the outcome of this PR is that delete == stop, does it just make sense to make stop an alias to delete in this PR?

thockin commented:
--cascade ?

On Thu, Apr 23, 2015 at 8:49 PM, Sam Ghods <notifications@github.com> wrote:

> Is recursive the right word here? Obviously it fits for RC's but doesn't
> this indicate whether to wait for a graceful stop of pods too? (Where
> recursive would be less appropriate.) Or does it not apply to pods?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7210#issuecomment-95791527>
> .
>

derekwaynecarr commented:
+1 for --cascade 

Sent from my iPhone

> On Apr 23, 2015, at 11:58 PM, Tim Hockin <notifications@github.com> wrote:
> 
> --cascade ?
> 
> On Thu, Apr 23, 2015 at 8:49 PM, Sam Ghods <notifications@github.com> wrote:
> 
> > Is recursive the right word here? Obviously it fits for RC's but doesn't
> > this indicate whether to wait for a graceful stop of pods too? (Where
> > recursive would be less appropriate.) Or does it not apply to pods?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/7210#issuecomment-95791527>
> > .
> >
> —
> Reply to this email directly or view it on GitHub.
>

ghodss commented:
Aren't options like "--delete-only=true" or "--graceful=false" more appropriate? --cascade still sounds like there are more objects to deal with, whereas that's not true with pods, or other objects we introduce that exclusive to themselves have a graceful and non-graceful way to stop.

thockin commented:
I think "graceful" and "cascade" are orthogonal ideas, and even though they
appear to be mutually exclusive now, we should not presume that they will
be forever.  Consider a world where deleting a pod with --cascade=false
might leave the pod's data volumes alive, while --cascade=true would
destroy them.  That's a very different concern than whether the pod gets 30
seconds of warning or not.

On Thu, Apr 23, 2015 at 11:31 PM, Sam Ghods <notifications@github.com>
wrote:

> Aren't options like "--delete-only=true" or "--graceful=false" more
> appropriate? --cascade still sounds like there are more objects to deal
> with, whereas that's not true with pods, or other objects we introduce that
> exclusive to themselves have a graceful and non-graceful way to stop.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7210#issuecomment-95817834>
> .
>

brendandburns commented:
Switched to ```--cascade```, @liggitt I added code so that if the flag is unset, and there is no reaper, then we skip back to the existing delete code.

I will add the alias in a different PR.

Please take another look.

Thanks!
--brendan

bgrant0607 commented on pkg/kubectl/cmd/delete_test.go:
```diff
@@ -50,6 +50,7 @@ func TestDeleteObjectByTuple(t *testing.T) {
 
 	cmd := NewCmdDelete(f, buf)
 	cmd.Flags().Set("namespace", "test")
+	cmd.Flags().Set("recursive", "false")
```
Should be change to cascade (everywhere).

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -95,8 +96,44 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+	if cmdutil.GetFlagBool(cmd, "cascade") {
+		err := ReapResult(r, f, out)
+		// If the error is something other than "not found" or the user explicitly requested "cascade" error out.
+		if !kubectl.IsNoSuchReaperError(err) || cmd.Flags().Lookup("cascade").Changed {
```
Commands can act on multiple objects of multiple different kinds, and visitors may return multiple errors.

Shouldn't we ignore NoSuchReaper errors in ReapResult rather than here?

brendandburns replied:
done.

brendandburns commented:
Comments addressed, please re-check.

Thanks!
--brendan

brendandburns commented:
Friendly ping on this.

Thanks!
--brendan

jlowdermilk commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -95,20 +96,59 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+		reaper, err := f.Reaper(info.Mapping)
+		if err != nil {
+			// If the error is "not found" and the user didn't explicitly set "cascade" just delete.
+			if kubectl.IsNoSuchReaperError(err) && !cmd.Flags().Lookup("cascade").Changed {
```
This will crash (nil deref I think) if you call it from kubectl/cmd/stop, because stop does not define a "cascade" flag. This error handling should be moved into `RunDelete`.

brendandburns replied:
done.  (can't do it in ```RunDelete``` because I do want to Stop those things that support stop, and not delete everything just because something else can't be stopped)

jlowdermilk replied:
Checking the flag for nil is kind of an ugly hack... How about this: change `ReapResult` to take a boolean 'defaultToDelete' which if true, makes ReapResult do a standard delete on resources for which there is no reaper available. That way you don't have to pass in the command object, and you can do the check for an explicit --cascade in `RunDelete`.

brendandburns replied:
done.

brendandburns commented:
Comment addressed, please re-check.

--brendan

jlowdermilk commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -95,20 +96,59 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+		reaper, err := f.Reaper(info.Mapping)
+		if err != nil {
+			// If the error is "not found" and the user didn't explicitly set "cascade" just delete.
+			if kubectl.IsNoSuchReaperError(err) && cmd.Flags().Lookup("cascade") != nil && !cmd.Flags().Lookup("cascade").Changed {
```
Repeating since it may have been lost in outdated diff

Checking the flag for nil is kind of an ugly hack... How about this: change ReapResult to take a boolean `defaultToDelete` instead of passing the Command object. If `defaultToDelete` is true, do a standard delete on resources for which there is no reaper available. That way you can move the check for an explicit `--cascade` into RunDelete, and we're not checking for nonexistent flags as a proxy for "which toplevel command is calling me".

brendandburns commented:
rebased.

jlowdermilk commented:
There's an unresolved comment in the diff for pkg/kubectl/cmd/delete.go that seems to be getting attached to the outdated diff thread. Please take a look and let me know if you can't see it or object to suggested change.

jlowdermilk commented on pkg/kubectl/cmd/stop.go:
```diff
@@ -86,3 +63,24 @@ func NewCmdStop(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	if r.Err() != nil {
+		return r.Err()
+	}
+	return ReapResult(r, f, out, true)
```
Shouldn't this be false? By running `stop` user is explicitly requesting reaper.

brendandburns replied:
done.

jlowdermilk commented:
lgtm aside from isDeleteDefault=false for RunStop.

brendandburns commented:
comment addressed, ptal.  unit tests are manually determined to be clean.

jlowdermilk commented:
lgtm
